grpc/transport-auth.go

```diff
@@ -0,0 +1,70 @@
+package grpc
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/credentials"
+)
+
+// MultiNameAuthenticator does a thing
+type MultiNameAuthenticator struct {
+	names map[string]*tls.Config
+}
+
+// NewMultiNameTLS also does a thing
+func NewMultiNameTLS(names map[string]*tls.Config) credentials.TransportAuthenticator {
+	return &MultiNameAuthenticator{names}
+}
+
+func (mna *MultiNameAuthenticator) ClientHandshake(addr string, rawConn net.Conn, timeout time.Duration) (net.Conn, credentials.AuthInfo, error) {
+	var errChannel chan error
+	if timeout != 0 {
+		errChannel = make(chan error, 2)
+		time.AfterFunc(timeout, func() {
+			errChannel <- errors.New("boulder/grpc: Dial timed out")
+		})
+	}
+	config, present := mna.names[addr]
```

Comment:
I believe concurrent reads from maps are safe and `tls.Config` is designed to be safe for concurrent use (nothing is supposed to mutate it).